hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/utils/ResourceLimitCache.java

```diff
@@ -79,7 +79,11 @@ public V remove(K key) {
   @Override
   public void removeIf(Predicate<K> predicate) {
     Objects.requireNonNull(predicate);
-    map.keySet().removeIf(predicate);
+    for (K key : new ArrayList<>(map.keySet())) {
```

Comment:
We can use iterator here to replace ArrayList.

Reply:
Thanks @lokeshj1703 for the review.

We could loop over `map.keySet()` directly, the `for-each` loop (`:`) uses iterator behind the scenes either way.  I wrapped key set in the list to avoid concurrent modification issues.

Reply:
I think we can use a combination of forEach and remove mehods exposed by ConcurrentMap interface. Since ConcurrentMap operations are thread safe, it should not throw concurrent modification errors.
Ref: https://stackoverflow.com/questions/3768554/is-iterating-concurrenthashmap-values-thread-safe

Reply:
Yes, `ConcurrentHashMap` doc says that its iterator does not throw `ConcurrentModificationException`.  However, `ResourceLimitCache` accepts any `ConcurrentMap`, and this interface does not state the same (i.e. I am not sure it is a requirement for all implementations).  So I preferred to be on the safe side.

Since `clear()` already iterates `keySet` directly, I guess we can still get rid of the wrapping `ArrayList`, so I have updated the patch.

Reply:
It might be safe to just skip the array copy here? As we already did in `clear()` below.

Reply:
I agree. Maybe we should replace the for loop with `ConcurrentMap.forEach(BiConsumer<? super K, ? super V> action)`. Since the method is exposed by the interface it should provide the thread safety guarantees.

Reply:
`ConcurrentMap#forEach` doesn't throw `ConcurrentModificationException` like `Map#forEach` does. I think it might be fine.